### DISCUSSION TITLE
Powershell Avalonia

### DISCUSSION THREAD

**Discussion by Hexality** [AvaloniaUI/Avalonia#10791] [Q&A]
Hi, does Avalonia support creation of UIs using c# assemblies directly on PowerShell like its possible with both WPF and WinForms?

![image](https://user-images.githubusercontent.com/17398632/227695581-f82f8689-6767-4746-b4f1-ff53dd5f3cb8.png)

**Comment by rstm-sf** [comment]
https://github.com/ironmansoftware/psavalonia

**Reply by Hexality** [reply]
PSsvalonia isn't really what I want due to how limited it is
you can't do advanced stuff with it like you can do while using wpf directly on powershell.

**Reply by Hexality** [reply]
PSAvalonia uses raw XAML, which isn't something I like to do due to how XAML works and behaves.

**Reply by ALIENQuake** [reply]
@Hexality Can you be more specific about the PSAvalonia/raw xaml limitations?

**Reply by Hexality** [reply]
Manually determining sets can be a pain, and XAML doesn't really integrate well as the xaml intellisense doesn't really work for xaml if embedded.
(Get-Content -Raw) does the job but defeats the purpose of having everything in a single file (which would be the optimal for a script).

**Reply by Hexality** [reply]
With "doing in file" I mean doing like I would with WPF
```ps1
using namespace System.Windows
Add-Type -Assembly PresentationFramwork
$mainForm = [Window]@{
    Width = 800
    Height = 480
}

$mainGrid = [Grid]@{}

$label = [Label]@{ 
    Content = "Hello World"
 }
$mainGrid.AddChild($label)
$mainForm.AddChild($mainGrid)
[void]$mainForm.ShowDialog()
```
(don't know if I wrote it right as I typed from my phone and can't test rn, but this is an example of what I mean).

**Comment by ALIENQuake** [comment]
@Hexality If you want to support for Avalonia 11.0.0-preview use this one https://github.com/smartguy1196/psavalonia

**Reply by Hexality** [reply]
https://github.com/AvaloniaUI/Avalonia/discussions/10791#discussioncomment-5429887

**Comment by awakecoding** [comment]
+1 I am also looking for a version of PSAvalonia that works with the latest version of Avalonia UI. Dynamic in-memory XAML loading would be important for scripted GUIs in PowerShell, files wouldn't be practical here

**Comment by mdgrs-mei** [comment]
This is now possible by using this module:

https://github.com/mdgrs-mei/GliderUI

On macOS, the GUI thread needs to be the main thread in a process but the thread where scripts run in PowerShell is not the main thread. I think the IPC approach that the module uses is actually the only way to create GUIs from PowerShell on macOS.
